common/datavalues2/src/data_value.rs

```diff
@@ -0,0 +1,318 @@
+        matches!(self, DataValue::Null)
+    }
+
+    pub fn value_type(&self) -> ValueType {
+        match self {
+            DataValue::Null => ValueType::Null,
+            DataValue::Boolean(_) => ValueType::Boolean,
+            DataValue::Int64(_) => ValueType::Int64,
+            DataValue::UInt64(_) => ValueType::UInt64,
+            DataValue::Float64(_) => ValueType::Float64,
+            DataValue::String(_) => ValueType::String,
+            DataValue::Array(_) => ValueType::Array,
+            DataValue::Struct(_) => ValueType::Struct,
+        }
+    }
+    // convert to minialized data type
+    pub fn data_type(&self) -> DataTypePtr {
+        match self {
+            DataValue::Null => Arc::new(NullType {}),
+            DataValue::Boolean(_) => BooleanType::arc(),
+            DataValue::Int64(n) => {
+                if *n >= i8::MIN as i64 && *n <= i8::MAX as i64 {
+                    return Int8Type::arc();
+                }
+                if *n >= i16::MIN as i64 && *n <= i16::MAX as i64 {
+                    return Int16Type::arc();
+                }
+                if *n >= i32::MIN as i64 && *n <= i32::MAX as i64 {
+                    return Int32Type::arc();
+                }
+                Int64Type::arc()
+            }
+            DataValue::UInt64(n) => {
+                if *n <= u8::MIN as u64 {
+                    return UInt8Type::arc();
+                }
+                if *n <= u16::MIN as u64 {
+                    return UInt16Type::arc();
+                }
+                if *n <= u32::MIN as u64 {
+                    return UInt32Type::arc();
+                }
+                UInt64Type::arc()
+            }
+            DataValue::Float64(_) => Float64Type::arc(),
+            DataValue::String(_) => StringType::arc(),
+            DataValue::Array(x) => {
+                let inner_type = if x.is_empty() {
+                    UInt8Type::arc()
+                } else {
+                    x[0].data_type()
+                };
+                Arc::new(ArrayType::create(inner_type))
+            }
+            DataValue::Struct(x) => {
+                let names = (0..x.len()).map(|i| format!("{}", i)).collect::<Vec<_>>();
+                let types = x.iter().map(|v| v.data_type()).collect::<Vec<_>>();
+                Arc::new(StructType::create(names, types))
+            }
+        }
+    }
+
+    #[inline]
+    pub fn is_integer(&self) -> bool {
+        matches!(self, |DataValue::Int64(_)| DataValue::UInt64(_))
+    }
+
+    #[inline]
+    pub fn is_signed_integer(&self) -> bool {
+        matches!(self, DataValue::Int64(_))
+    }
+
+    #[inline]
+    pub fn is_unsigned_integer(&self) -> bool {
+        matches!(
+            self,
+                | DataValue::UInt64(_)
+        )
+    }
+
+    pub fn to_values(&self, size: usize) -> Result<Vec<DataValue>> {
+        Ok((0..size).map(|_| self.clone()).collect())
+    }
+
+    pub fn as_u64(&self) -> Result<u64> {
```

Comment:
Comparing to `as_xxx`, they behave more like a `to_xxx` function. Does the names have their own meaning?

Reply:
if we use `to_xxx`, we must consume `self` by clippy.